### PR TITLE
7903181: port jextract for foreign-preview changes

### PR DESCRIPTION
**Pull request by sundararajana**
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed (no reviews required)

### Issue
 * [CODETOOLS-7903181](https://bugs.openjdk.java.net/browse/CODETOOLS-7903181): port jextract for foreign-preview changes

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.java.net/census#mcimadamore) (@mcimadamore - Committer) ⚠️ Review applies to [00e820c2](https://git.openjdk.java.net/jextract/pull/35/files/00e820c27587f6e9abb59b31c5a8797efaef78f4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jextract pull/35/head:pull/35` \
`$ git checkout pull/35`

Update a local copy of the PR: \
`$ git checkout pull/35` \
`$ git pull https://git.openjdk.java.net/jextract pull/35/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 35`

View PR using the GUI difftool: \
`$ git pr show -t 35`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jextract/pull/35.diff">https://git.openjdk.java.net/jextract/pull/35.diff</a>

</details>

### CODE REVIEW

File: samples/cblas/TestBlas.java [hunk 1]
```diff
@@ -54,18 +54,17 @@ public static void main(String[] args) {
         alpha = 1;
         beta = 0;
 
-        try (var scope = ResourceScope.newConfinedScope()) {
-            var allocator = SegmentAllocator.newNativeArena(scope);
-            var a = allocator.allocateArray(C_DOUBLE, new double[] {
+        try (var session = MemorySession.openConfined()) {
+            var a = session.allocateArray(C_DOUBLE, new double[] {
```
Comment (mcimadamore) [inline samples/cblas/TestBlas.java:58]: Note: this could use varargs (maybe for later)

File: samples/dlopen/Dlopen.java [hunk 2]
```diff
@@ -30,46 +30,44 @@
...
 
 public class Dlopen {
-    // implementation of SymbolLookup that loads a given shared object using dlopen
+    // implementation of Symbol lookup that loads a given shared object using dlopen
     // and looks up symbols using dlsym
-    private static SymbolLookup lookup(String libraryName, ResourceScope scope) {
-        try (ResourceScope openScope = ResourceScope.newConfinedScope()) {
-            var openScopeAllocator = SegmentAllocator.nativeAllocator(openScope);
-            var handle = dlopen(openScopeAllocator.allocateUtf8String(libraryName), RTLD_LOCAL());
+    private static Function<String, Optional<MemorySegment>> lookup(String libraryName, MemorySession session) {
```
Comment (mcimadamore) [inline samples/dlopen/Dlopen.java:42]: Does this sample still make sense with the new `libraryLookup` ?

File: samples/lapack/TestLapack.java [hunk 3]
```diff
@@ -29,22 +29,21 @@
...
 import static lapack.lapacke_h.*;
 
 public class TestLapack {
     public static void main(String[] args) {
 
         /* Locals */
-        try (var scope = ResourceScope.newConfinedScope()) {
-            var allocator = SegmentAllocator.newNativeArena(scope);
-            var A = allocator.allocateArray(C_DOUBLE, new double[]{
+        try (var session = MemorySession.openConfined()) {
```
Comment (mcimadamore) [inline samples/lapack/TestLapack.java:42]: Note: this sample could use varargs (maybe for later)

File: samples/libffmpeg/LibffmpegMain.java [hunk 4]
```diff
@@ -32,14 +32,14 @@
...
-import jdk.incubator.foreign.*;
+import java.lang.foreign.*;
 import libffmpeg.AVCodecContext;
 import libffmpeg.AVFormatContext;
 import libffmpeg.AVFrame;
 import libffmpeg.AVPacket;
 import libffmpeg.AVStream;
 import static libffmpeg.Libffmpeg.*;
-import static jdk.incubator.foreign.MemoryAddress.*;
+import static java.lang.foreign.MemoryAddress.*;
```
Comment (mcimadamore) [inline samples/libffmpeg/LibffmpegMain.java:42]: Most of other samples statically import just NULL

File: src/main/java/org/openjdk/jextract/clang/libclang/CXString.java [hunk 5]
```diff
@@ -47,13 +47,13 @@ public class CXString {
...
     public static MemoryAddress data$get(MemorySegment seg) {
-        return (jdk.incubator.foreign.MemoryAddress)CXString.data$VH.get(seg);
+        return (java.lang.foreign.MemoryAddress)CXString.data$VH.get(seg);
     }
     public static void data$set( MemorySegment seg, MemoryAddress x) {
         CXString.data$VH.set(seg, x);
     }
     public static MemoryAddress data$get(MemorySegment seg, long index) {
-        return (jdk.incubator.foreign.MemoryAddress)CXString.data$VH.get(seg.asSlice(index*sizeof()));
+        return (java.lang.foreign.MemoryAddress)CXString.data$VH.get(seg.asSlice(index*sizeof()));
```
Comment (mcimadamore) [inline src/main/java/org/openjdk/jextract/clang/libclang/CXString.java:56]: Maybe jextract should not emit a fully qualified name for the cast target? There's an import anyways.

File: src/main/java/org/openjdk/jextract/impl/StructBuilder.java [hunk 6]
```diff
@@ -254,33 +254,11 @@ private void emitAllocatorAllocateArray() {
...
-        append("}\n");
-        decrAlign();
-    }
-
     private void emitOfAddressScoped() {
         incrAlign();
         indent();
         append(MEMBER_MODS);
-        append(" MemorySegment ofAddress(MemoryAddress addr, ResourceScope scope) { return RuntimeHelper.asArray(addr, $LAYOUT(), 1, scope); }\n");
+        append(" MemorySegment ofAddress(MemoryAddress addr, MemorySession scope) { return RuntimeHelper.asArray(addr, $LAYOUT(), 1, scope); }\n");
```
Comment (mcimadamore) [inline src/main/java/org/openjdk/jextract/impl/StructBuilder.java:261]: s/scope/session

File: src/main/resources/org/openjdk/jextract/impl/resources/RuntimeHelper.java.template [hunk 7]
```diff
@@ -71,28 +70,28 @@ final class RuntimeHelper {
                 orElse(null);
     }
 
-    static final <Z> NativeSymbol upcallStub(Class<Z> fi, Z z, FunctionDescriptor fdesc, ResourceScope scope) {
+    static final <Z> MemorySegment upcallStub(Class<Z> fi, Z z, FunctionDescriptor fdesc, MemorySession scope) {
```
Comment (mcimadamore) [inline src/main/resources/org/openjdk/jextract/impl/resources/RuntimeHelper.java.template:73]: s/scope/session

File: src/main/resources/org/openjdk/jextract/impl/resources/RuntimeHelper.java.template [hunk 8]
```diff
@@ -71,28 +70,28 @@ final class RuntimeHelper {
...
+            MethodHandle handle = MH_LOOKUP.findVirtual(fi, "apply", Linker.upcallType(fdesc));
             handle = handle.bindTo(z);
             return LINKER.upcallStub(handle, fdesc, scope);
         } catch (Throwable ex) {
             throw new AssertionError(ex);
         }
     }
 
-    static MemorySegment asArray(MemoryAddress addr, MemoryLayout layout, int numElements, ResourceScope scope) {
+    static MemorySegment asArray(MemoryAddress addr, MemoryLayout layout, int numElements, MemorySession scope) {
```
Comment (mcimadamore) [inline src/main/resources/org/openjdk/jextract/impl/resources/RuntimeHelper.java.template:83]: s/scope/session

File: test/jtreg/generator/test8244938/Test8244938.java [hunk 9]
```diff
@@ -47,7 +47,7 @@
 public class Test8244938 {
     @Test
     public void testNestedStructReturn() {
-         try (ResourceScope scope = ResourceScope.newConfinedScope()) {
+         try (MemorySession scope = MemorySession.openConfined()) {
```
Comment (mcimadamore) [inline test/jtreg/generator/test8244938/Test8244938.java:50]: s/scope/session

File: test/jtreg/generator/test8246341/LibTest8246341Test.java [hunk 10]
```diff
@@ -21,16 +21,16 @@
...
+import java.lang.foreign.MemorySegment;
+import java.lang.foreign.MemorySession;
+import java.lang.foreign.SegmentAllocator;
 import org.testng.annotations.Test;
 import test.jextract.test8246341.*;
 import static org.testng.Assert.assertEquals;
 import static org.testng.Assert.assertTrue;
 import static test.jextract.test8246341.test8246341_h.*;
-import static jdk.incubator.foreign.CLinker.*;
+import static java.lang.foreign.Linker.*;
```
Comment (mcimadamore) [inline test/jtreg/generator/test8246341/LibTest8246341Test.java:33]: is this static import needed?

File: test/jtreg/generator/test8246341/LibTest8246341Test.java [hunk 11]
```diff
@@ -52,7 +52,7 @@ public class LibTest8246341Test {
     @Test
     public void testPointerArray() {
         boolean[] callbackCalled = new boolean[1];
-        try (ResourceScope scope = ResourceScope.newConfinedScope()) {
+        try (MemorySession scope = MemorySession.openConfined()) {
```
Comment (mcimadamore) [inline test/jtreg/generator/test8246341/LibTest8246341Test.java:55]: s/scope/session

File: test/jtreg/generator/test8246341/LibTest8246341Test.java [hunk 12]
```diff
@@ -69,7 +69,7 @@ public void testPointerArray() {
 
     @Test
     public void testPointerAllocate() {
-        try (var scope = ResourceScope.newConfinedScope()) {
+        try (var scope = MemorySession.openConfined()) {
```
Comment (mcimadamore) [inline test/jtreg/generator/test8246341/LibTest8246341Test.java:72]: s/scope/session

File: test/jtreg/generator/test8246400/LibTest8246400Test.java [hunk 13]
```diff
@@ -51,7 +51,7 @@ public class LibTest8246400Test {
     @Test
     public void testSegmentRegister() {
         MemorySegment sum = null;
-        try (ResourceScope scope = ResourceScope.newConfinedScope()) {
+        try (MemorySession scope = MemorySession.openConfined()) {
```
Comment (mcimadamore) [inline test/jtreg/generator/test8246400/LibTest8246400Test.java:54]: s/scope/session

File: test/jtreg/generator/test8252016/Test8252016.java [hunk 14]
```diff
@@ -21,16 +21,16 @@
...
+import java.lang.foreign.VaList;
 import org.testng.annotations.Test;
 
-import jdk.incubator.foreign.MemorySegment;
+import java.lang.foreign.MemorySegment;
 
 import static org.testng.Assert.assertEquals;
 import static test.jextract.vsprintf.vsprintf_h.*;
-import static jdk.incubator.foreign.CLinker.*;
+import static java.lang.foreign.Linker.*;
```
Comment (mcimadamore) [inline test/jtreg/generator/test8252016/Test8252016.java:33]: redundant? E.g. we only need to import VaList

File: test/jtreg/generator/test8254983/LibTest8254983Test.java [hunk 15]
```diff
@@ -48,7 +48,7 @@
 public class LibTest8254983Test {
     @Test
     public void testOuterStruct() {
-        try (ResourceScope scope = ResourceScope.newConfinedScope()) {
+        try (MemorySession scope = MemorySession.openConfined()) {
```
Comment (mcimadamore) [inline test/jtreg/generator/test8254983/LibTest8254983Test.java:51]: s/scope/session

File: test/jtreg/generator/test8254983/LibTest8254983Test.java [hunk 16]
```diff
@@ -59,7 +59,7 @@ public void testOuterStruct() {
     @Test
     public void testInnerStruct() {
         assertEquals(((GroupLayout)Foo._union._struct.$LAYOUT()).memberLayouts().size(), 2);
-        try (ResourceScope scope = ResourceScope.newConfinedScope()) {
+        try (MemorySession scope = MemorySession.openConfined()) {
```
Comment (mcimadamore) [inline test/jtreg/generator/test8254983/LibTest8254983Test.java:62]: s/scope/session

File: test/jtreg/generator/test8257892/LibUnsupportedTest.java [hunk 17]
```diff
@@ -64,7 +64,7 @@ public void testAllocateFoo() {
 
     @Test
     public void testGetFoo() {
-        try (ResourceScope scope = ResourceScope.newConfinedScope()) {
+        try (MemorySession scope = MemorySession.openConfined()) {
```
Comment (mcimadamore) [inline test/jtreg/generator/test8257892/LibUnsupportedTest.java:67]: s/scope/session

File: test/jtreg/generator/test8258605/LibTest8258605Test.java [hunk 18]
```diff
@@ -48,7 +48,7 @@
 public class LibTest8258605Test {
     @Test
     public void testFunctionCallback() {
-        try (ResourceScope scope = ResourceScope.newConfinedScope()) {
+        try (MemorySession scope = MemorySession.openConfined()) {
```
Comment (mcimadamore) [inline test/jtreg/generator/test8258605/LibTest8258605Test.java:51]: s/scope/session

File: test/jtreg/generator/test8261511/Test8261511.java [hunk 19]
```diff
@@ -46,7 +46,7 @@
 public class Test8261511 {
     @Test
     public void test() {
-        try (ResourceScope scope = ResourceScope.newConfinedScope()) {
+        try (MemorySession scope = MemorySession.openConfined()) {
```
Comment (mcimadamore) [inline test/jtreg/generator/test8261511/Test8261511.java:49]: s/scope/session

File: test/jtreg/generator/testFunctionPointer/LibFuncPtrTest.java [hunk 20]
```diff
@@ -43,7 +43,7 @@
 public class LibFuncPtrTest {
     @Test
     public void test() {
-        try (ResourceScope scope = ResourceScope.newConfinedScope()) {
+        try (MemorySession scope = MemorySession.openConfined()) {
```
Comment (mcimadamore) [inline test/jtreg/generator/testFunctionPointer/LibFuncPtrTest.java:46]: s/scope/session

File: test/jtreg/generator/testStruct/LibStructTest.java [hunk 21]
```diff
@@ -50,7 +49,7 @@
 public class LibStructTest {
     @Test
     public void testMakePoint() {
-        try (ResourceScope scope = ResourceScope.newConfinedScope()) {
+        try (MemorySession scope = MemorySession.openConfined()) {
```
Comment (mcimadamore) [inline test/jtreg/generator/testStruct/LibStructTest.java:52]: s/scope/session

File: test/jtreg/generator/testStruct/LibStructTest.java [hunk 22]
```diff
@@ -59,7 +58,7 @@ public void testMakePoint() {
 
     @Test
     public void testAllocate() {
-        try (ResourceScope scope = ResourceScope.newConfinedScope()) {
+        try (MemorySession scope = MemorySession.openConfined()) {
```
Comment (mcimadamore) [inline test/jtreg/generator/testStruct/LibStructTest.java:61]: s/scope/session

File: test/jtreg/generator/testStruct/LibStructTest.java [hunk 23]
```diff
@@ -70,7 +69,7 @@ public void testAllocate() {
 
     @Test
     public void testAllocateArray() {
-        try (ResourceScope scope = ResourceScope.newConfinedScope()) {
+        try (MemorySession scope = MemorySession.openConfined()) {
```
Comment (mcimadamore) [inline test/jtreg/generator/testStruct/LibStructTest.java:72]: s/scope/session

File: test/testng/org/openjdk/jextract/test/toolprovider/TestClassGeneration.java [hunk 24]
```diff
@@ -181,7 +180,7 @@ public void testStructMember(String structName, MemoryLayout memberLayout, Class
         Class<?> structCls = loader.loadClass("com.acme." + structName);
         Method layout_getter = checkMethod(structCls, "$LAYOUT", MemoryLayout.class);
         MemoryLayout structLayout = (MemoryLayout) layout_getter.invoke(null);
-        try (ResourceScope scope = ResourceScope.newConfinedScope()) {
+        try (MemorySession scope = MemorySession.openConfined()) {
```
Comment (mcimadamore) [inline test/testng/org/openjdk/jextract/test/toolprovider/TestClassGeneration.java:183]: s/scope/session

File: test/testng/org/openjdk/jextract/test/toolprovider/TestNested.java [hunk 25]
```diff
@@ -155,7 +155,7 @@ private static void checkAccessors(Class<?> cls, MemoryLayout layout, String fie
             if (type == MemorySegment.class) {
                 Method slicer = cls.getMethod(fieldName + "$slice", MemorySegment.class);
                 assertEquals(slicer.getReturnType(), MemorySegment.class);
-                try (ResourceScope scope = ResourceScope.newConfinedScope()) {
+                try (MemorySession scope = MemorySession.openConfined()) {
```
Comment (mcimadamore) [inline test/testng/org/openjdk/jextract/test/toolprovider/TestNested.java:158]: s/scope/session

File: test/testng/org/openjdk/jextract/test/toolprovider/TestNested.java [hunk 26]
```diff
@@ -167,7 +167,7 @@ private static void checkAccessors(Class<?> cls, MemoryLayout layout, String fie
                 assertEquals(setter.getReturnType(), void.class);
 
                 Object zero = MethodHandles.zero(type).invoke();
-                try (ResourceScope scope = ResourceScope.newConfinedScope()) {
+                try (MemorySession scope = MemorySession.openConfined()) {
```
Comment (mcimadamore) [inline test/testng/org/openjdk/jextract/test/toolprovider/TestNested.java:170]: s/scope/session

File: samples/dlopen/Dlopen.java (hunk 2 above)
Comment (sundararajana) [reply samples/dlopen/Dlopen.java:42]: May be not. But as a POSIX specific example of dlopen/dlsym/dladdr etc. We may be expand to use flags, dladdr etc.

File: src/main/java/org/openjdk/jextract/clang/libclang/CXString.java (hunk 5 above)
Comment (sundararajana) [reply src/main/java/org/openjdk/jextract/clang/libclang/CXString.java:56]: Yes, pre-existing issue I think. Will address it in another bug

File: test/jtreg/generator/testStruct/LibStructTest.java (hunk 23 above)
Comment (sundararajana) [reply test/jtreg/generator/testStruct/LibStructTest.java:72]: Will fix it.

File: test/testng/org/openjdk/jextract/test/toolprovider/TestNested.java (hunk 26 above)
Comment (sundararajana) [reply test/testng/org/openjdk/jextract/test/toolprovider/TestNested.java:170]: will fix it

File: test/testng/org/openjdk/jextract/test/toolprovider/TestClassGeneration.java (hunk 24 above)
Comment (sundararajana) [reply test/testng/org/openjdk/jextract/test/toolprovider/TestClassGeneration.java:183]: will fix it.

File: samples/cblas/TestBlas.java (hunk 1 above)
Comment (sundararajana) [reply samples/cblas/TestBlas.java:58]: will use varargs

**Review by mcimadamore — APPROVED** [review]

File: samples/libffmpeg/LibffmpegMain.java (hunk 4 above)
Comment (sundararajana) [reply samples/libffmpeg/LibffmpegMain.java:42]: Will fix with specific import

**Review by mcimadamore — APPROVED** [review]

**Pull request closed without merging**